src/component.jl

```diff
@@ -184,29 +195,120 @@ Base.sortperm(c::SharedComponent) = sortperm(c.data)
     end
     return h
 end
+
+#TODO: These boundchecks etc are to ensure in the case of || in @entities_in nobody is trying to load some
+#      invalid memory. Do we really need ||?
+@generated function Base.getindex(t::Tup, ::Type{T}) where {Tup <: Tuple, T<:ComponentData}
+    id = findfirst(x -> eltype(x) == T, Tup.parameters)
+    # return :(unsafe_load(@inbounds getindex(t, $id))::T) # without ||
+    return quote
+        $(Expr(:meta, :inline))
+        $(Expr(:meta, :propagate_inbounds))
+        ptr = getindex(t, $id)
+        @boundscheck if ptr === Ptr{T}()
+            throw(BoundsError(t, T))
+        end
+        unsafe_load(ptr)::T
+    end
+end
+
+@generated function Base.setindex!(t::Tup, x::T, ::Type{T}) where {Tup <: Tuple, T<:ComponentData}
+    id = findfirst(x -> eltype(x) == T, Tup.parameters)
+    # return :(unsafe_store!(@inbounds getindex(t, $id), x)) # without ||
+    return quote
+        $(Expr(:meta, :inline))
+        $(Expr(:meta, :propagate_inbounds))
+        ptr = getindex(t, $id)
+        @boundscheck if ptr === Ptr{T}()
+            throw(BoundsError(t, T))
+        end
+        unsafe_store!(ptr, x)
+    end
+end
 ########################################
 #                                      #
 #            Iteration                 #
 #                                      #
 ########################################
-struct EntityIterator{T <: Union{IndicesIterator,Indices,AbstractGroup}}
+struct EntityIterator{T <: Union{IndicesIterator,Indices,AbstractGroup}, TT <: Tuple}
     it::T
+    components::TT
 end
 
-Base.eltype(::EntityIterator) = Entity
+Base.eltype(::EntityIterator{T, TT}) where {T, TT} = EntityState{Tuple{map(x->Ptr{eltype(x)}, TT.parameters)...}}
 Base.IteratorSize(i::EntityIterator) = Base.IteratorSize(i.it)
 Base.length(i::EntityIterator) = length(i.it)
 
-function Base.iterate(i::EntityIterator, state = 1)
+struct EntityState{TT<:Tuple} <: AbstractEntity
+    e::Entity
+    pointers::TT
+end
+
+@generated function Base.getproperty(e::EntityState{TT}, f::Symbol) where {TT}
+    fn_to_DT = Dict{Symbol, DataType}()
+    ex = :(getfield(e, f))
+    ex = Expr(:elseif, :(f === :id), :(return getfield(getfield(e, :e), :id)::Int), ex)
+    for PDT in TT.parameters
+        DT = eltype(PDT)
+        for (fn, ft) in zip(fieldnames(DT), fieldtypes(DT))
+            if haskey(fn_to_DT, fn)
+                fnq = QuoteNode(fn)
+                DT_ = fn_to_DT[fn]
+                ft_ = fieldtype(DT_, fn)
+                ex = MacroTools.postwalk(ex) do x
+                    if @capture(x, return getfield(getfield(e, :pointers)[$DT_], $fnq)::$ft_)
```

Comment:
Since you're constructing `ex` here yourself, you don't need `MacroTools` to deconstruct the expression once again.

Instead, just keep track of the current list of field names in a `Set` as you go through the loop?

Reply:
Yea not clear to me what the hell my plan was here. 4 am mid sleep -> wake up -> idea -> implement -> works :white_check_mark: -> never check implementation again :sweat_smile:. 
Anyway, since I'll indeed just turn it into an error, this really doesn't matter anymore, indeed I already use the keys of the dict as the previously saved fn's.

Reply:
Hehe :)

Reply:
Oh now I remember, the point is that each previously created elseif with the ambiguous symbol needs to be replaced by the error.

Reply:
Oh, I see. In that case, a two-pass algorithm might be clearer.  First pass collects a `Set` of any non-unique names.  Second pass does the actual code generation.